### ISSUE TITLE
Filebeat 7.4 Kubernetes Autodiscovery Memory Leak

### ISSUE THREAD

**Issue by lystor** [elastic/beats#14189]
Hello

We have got memory leak on Filebeat 7.4.0 after upgrade from 7.0.0 on all kubernetes nodes (12x).

- Version: 7.4.0
- Operating System: Official Docker Image
- Steps to Reproduce: 

```
# helm upgrade filebeat-nova stable/filebeat --install \
        --namespace=kube-system --values=values_override.yaml --version=3.1.0
```

values_override.yaml:
```
data:
  hostPath: "/var/lib/filebeat-nova"

image:
  tag: "7.4.0"

monitoring:
  enabled: false

privileged: true

overrideConfig:
  filebeat.autodiscover:
    providers:
      - type: kubernetes
        templates:
          - config:
              - type: container
                clean_inactive: "168h"
                clean_removed: true
                close_eof: false
                close_inactive: "5m"
                close_renamed: true
                close_removed: true
                close_timeout: "25h"
                exclude_lines:
                  - "Using deprecated option"
                ignore_older: "24h"
                max_bytes: 1048576
                paths:
                  - "/var/log/containers/*-${data.kubernetes.container.id}.log"
                processors:
                  - add_kubernetes_metadata:
                      in_cluster: true
                  - decode_json_fields:
                      fields:
                        - "message"
                      max_depth: 1
                      process_array: false
                      overwrite_keys: true
                      target: ""
                scan_frequency: "10s"

  monitoring.enabled: true
  monitoring.elasticsearch:
    hosts:
      - "http://elk-nova-1.example.local:9200"
      - "http://elk-nova-2.example.local:9200"
      - "http://elk-nova-3.example.local:9200"

  output.elasticsearch:
    bulk_max_size: 10
    hosts:
      - "http://elk-nova-1.example.local:9200"
      - "http://elk-nova-2.example.local:9200"
      - "http://elk-nova-3.example.local:9200"
    indices:
      - index: "filebeat-game01-%{[kubernetes.container.name]}-%{+xxxx.ww}"
        when:
          and:
            - contains:
                kubernetes.labels.type: "game01"
            - not.regexp:
                kubernetes.container.name: "^(istio-proxy|redis)$"
      - index: "filebeat-lds-%{+xxxx.ww}"
        when.equals:
          kubernetes.namespace: "lds"
      - index: "filebeat-thirdparty-%{[kubernetes.container.name]}-%{+xxxx.ww}"
        when.has_fields:
          - "kubernetes.container.name"
    worker: 2

  setup.dashboards.enabled: false
  setup.ilm.enabled: false
  setup.template.enabled: false
```
![s1](https://user-images.githubusercontent.com/7275333/67300477-a9d5bd00-f4f6-11e9-8c9c-8b99b881c573.png)

![s2](https://user-images.githubusercontent.com/7275333/67300619-eb666800-f4f6-11e9-9705-41da1d3dad58.png)

![s3](https://user-images.githubusercontent.com/7275333/67300700-0df88100-f4f7-11e9-8ccf-22bdb584a5b7.png)

Please fix the bug.
Thank you

--
Mykola

**Comment by lystor** [comment]
Memory usage of Filebeat 7.4 with disabled kubernetes autodiscovery is about 5-10 MB:
```
data:
  hostPath: "/var/lib/filebeat-nova"

image:
  tag: "7.4.0"

monitoring:
  enabled: false

privileged: true

resources:
   limits:
      cpu: 100m
      memory: 1000Mi
   requests:
      cpu: 100m
      memory: 1000Mi

overrideConfig:
  filebeat.inputs:
    - type: container
      clean_inactive: "168h"
      clean_removed: true
      close_eof: false
      close_inactive: "5m"
      close_renamed: true
      close_removed: true
      close_timeout: "25h"
      exclude_lines:
        - "Using deprecated option"
      ignore_older: "24h"
      max_bytes: 1048576
      paths:
        - "/var/log/containers/*.log"
      scan_frequency: "10s"

  monitoring.enabled: true
  monitoring.elasticsearch:
    hosts:
      - "http://elk-nova-1.example.local:9200"
      - "http://elk-nova-2.example.local:9200"
      - "http://elk-nova-3.example.local:9200"

  output.elasticsearch:
    bulk_max_size: 10
    hosts:
      - "http://elk-nova-1.example.local:9200"
      - "http://elk-nova-2.example.local:9200"
      - "http://elk-nova-3.example.local:9200"
    indices:
      - index: "filebeat-game01-%{[kubernetes.container.name]}-%{+xxxx.ww}"
        when:
          and:
            - contains:
                kubernetes.labels.type: "game01"
            - not.regexp:
                kubernetes.container.name: "^(istio-proxy|redis)$"
      - index: "filebeat-lds-%{+xxxx.ww}"
        when.equals:
          kubernetes.namespace: "lds"
      - index: "filebeat-thirdparty-%{[kubernetes.container.name]}-%{+xxxx.ww}"
        when.has_fields:
          - "kubernetes.container.name"
      - index: "filebeat-unknown-%{+xxxx.ww}"
    worker: 2

  processors:
    - add_kubernetes_metadata:
        in_cluster: true
        matchers:
          - logs_path:
              logs_path: "/var/log/containers/"
    - decode_json_fields:
        fields:
          - "message"
        max_depth: 1
        process_array: false
        overwrite_keys: true
        target: ""

  setup.dashboards.enabled: false
  setup.ilm.enabled: false
  setup.template.enabled: false
```

**Comment by lystor** [comment]
Removal of add_kubernetes_metadata processor fixes the issue.

**Comment by jsoriano** [comment]
@lystor thanks for reporting! There was actually a memory leak in recent versions of Beats, it should be fixed by #14259.
